Add docs for GPU support (#660)
trixi-framework/TrixiParticles.jl: docs/src/gpu.md

# GPU Support

> We are currently actively working on GPU support.

Reviewer: ```suggestion
GPU support is still an experimental feature that is actively being worked on.
```

As of now, the [`WeaklyCompressibleSPHSystem`](@ref) and the [`BoundarySPHSystem`](@ref)
are supported on GPUs.
We have tested this on GPUs by Nvidia and AMD.

> To run a simulation on a GPU, we need to use the [`FullGridCellList`](@ref)
> as cell list inside the [`GridNeighborhoodSearch`](@ref).

Reviewer: ```suggestion
To run a simulation on a GPU, we need to use the [`FullGridCellList`](@ref)
as cell list for the used [`GridNeighborhoodSearch`](@ref).
```

This cell list requires a bounding box for the domain, unlike the default cell list, which
uses an unbounded domain.
For simulations that are bounded by a closed tank, we can use the boundary of the tank
to obtain the bounding box as follows.
```jldoctest gpu; output=false, setup=:(using TrixiParticles; trixi_include(@__MODULE__, joinpath(examples_dir(), "fluid", "hydrostatic_water_column_2d.jl"), sol=nothing))
search_radius = TrixiParticles.compact_support(smoothing_kernel, smoothing_length)
min_corner = minimum(tank.boundary.coordinates, dims=2) .- search_radius
max_corner = maximum(tank.boundary.coordinates, dims=2) .+ search_radius
cell_list = TrixiParticles.PointNeighbors.FullGridCellList(; min_corner, max_corner)

# output
PointNeighbors.FullGridCellList{PointNeighbors.DynamicVectorOfVectors{Int32, Matrix{Int32}, Vector{Int32}, Base.RefValue{Int32}}, Nothing, SVector{2, Float64}, SVector{2, Float64}}(Vector{Int32}[], nothing, [-0.24500000000000002, -0.24500000000000002], [1.245, 1.245])

> ```

Reviewer: Can't we default to setting this bounding box?

Author: Not really. What if we don't use a tank? Or if the tank doesn't have a closed top? 
I think it makes more sense to work on cell lists that support unbounded domains.

Reviewer: I am not saying to remove the option to overwrite the default bounding box. But using the minimum/maximum as a default seems to me to be a reasonable choice.

Author: The problem is that this is an option of the neighborhood search. The neighborhood search for fluid-fluid only sees the coordinates of the fluid, so we can't set the bounding box to the maximum and minimum of the boundary coordinates.

Reviewer: Yes it would need to be set in semidiscretization I guess.

Author: Discussion in https://github.com/trixi-framework/TrixiParticles.jl/issues/665.

to the [`Semidiscretization`](@ref).
```jldoctest gpu; output=false
semi = Semidiscretization(fluid_system, boundary_system,
                          neighborhood_search=GridNeighborhoodSearch{2}(; cell_list))

# output
┌──────────────────────────────────────────────────────────────────────────────────────────────────┐
│ Semidiscretization                                                                               │
│ ══════════════════                                                                               │
│ #spatial dimensions: ………………………… 2                                                                │
│ #systems: ……………………………………………………… 2                                                                │
│ neighborhood search: ………………………… GridNeighborhoodSearch                                           │
│ total #particles: ………………………………… 636                                                              │
└──────────────────────────────────────────────────────────────────────────────────────────────────┘
```

At this point, we should run the simulation and make sure that it still works and that

> the bounding box is large enough.

Reviewer: Can you give an example when the bounding box would not be large enough?

Reviewer: What I mean is to improve the wording so that you mention cases where a larger bounding box should be used like an open tank etc.

Then, we only need to specify the data type that is used for the simulation.
On an Nvidia GPU, we specify:
```julia
using CUDA
ode = semidiscretize(semi, tspan, data_type=CuArray)
```
On an AMD GPU, we use:
```julia
using AMDGPU
ode = semidiscretize(semi, tspan, data_type=ROCArray)
```
Then, we can run the simulation as usual.

> All data is transferred to the GPU during initialization and all loop over particles
> and neighbors are launched as GPU kernels.

Reviewer: ```suggestion
All data is transferred to the GPU during initialization and all loops over particles
and neighbors will be executed on the GPU.
```

Author: I think it's clearer for people if we mention GPU kernels. Or they will be asking "how is it executed on the GPU? Don't we need a kernel?"

Reviewer: How about this than:
All data is transferred to the GPU during initialization, and computations looping over particles and their neighbors are processed directly on the GPU, utilizing GPU kernels generated via KernelAbstractions.jl.

> Data is only copied to the CPU for saving VTK files with the [`SolutionSavingCallback`](@ref).

Reviewer: ```suggestion
Data is only copied to the CPU for saving VTK files via the [`SolutionSavingCallback`](@ref).
```

> Data is only copied to the CPU for saving VTK files with the [`SolutionSavingCallback`](@ref).

Reviewer: Probably postprocessing callback as well?

Author: I think that's not working yet.

At this point, we should run the simulation and make sure that it still works and that
the bounding box is large enough.

> For some simulations where particles move outside the initial tank coordiantes,

Reviewer: coordinates
